### PR TITLE
Some update to tr10 config

### PR DESCRIPTION
**Pull request by thomasw21**

### CODE REVIEW

File: train/tr10-13B-ml/tr10-13B.slurm [hunk 1]
```diff
@@ -46,7 +46,7 @@ GLOBAL_BATCH_SIZE=2048
 
 NLAYERS=40
 NHIDDEN=5120
-NHEADS=32
```
Comment (thomasw21) [inline train/tr10-13B-ml/tr10-13B.slurm:49]: I don't why we chose 32? We seem to have updated the `NHIDDEN` value to be 5120 because it was divisible by 128, and `5120 // 128 = 40`.

https://huggingface.slack.com/archives/C01NHER1JLS/p1627034738272600?thread_ts=1626827659.189400&cid=C01NHER1JLS

cc @VictorSanh @stas00 @mryab (People who were involved in the original post)

File: train/tr10-13B-ml/tr10-13B.slurm [hunk 2]
```diff
@@ -57,13 +57,14 @@ OPTIMIZER_ARGS=" \
...
-    --lr 6e-5 \
+    --lr 6e-4 \
     --min-lr 6e-6 \
     --lr-decay-style cosine \
-    --lr-decay-samples 126_953_125 \
     --lr-warmup-samples 216_320 \
     --clip-grad 1.0 \
     --weight-decay 1e-1 \
+    --hidden-dropout 0.0 \
+    --attention-dropout 0.0 \
```
Comment (thomasw21) [inline train/tr10-13B-ml/tr10-13B.slurm:67]: https://arxiv.org/abs/2010.11934 showed strong performance loss when using dropout (table 4). Though it was enc/dec architecture, there's probably no reason that it would benefit our dec only arch. We are currently evaluating this on 1B3 scale. https://huggingface.co/bigscience/tr3o-1B3-pile-no-dropout-logs

File: train/tr10-13B-ml/tr10-13B.slurm [hunk 3]
```diff
@@ -80,7 +81,7 @@ GPT_ARGS=" \
     --micro-batch-size $MICRO_BATCH_SIZE \
     --rampup-batch-size 16 16 6_000_000 \
     --global-batch-size $GLOBAL_BATCH_SIZE \
-    --train-samples 300_000_000 \
+    --train-samples $((3000000000 / $SEQ_LEN + 1)) \
```
Comment (thomasw21) [inline train/tr10-13B-ml/tr10-13B.slurm:84]: This should be `1464844` right as all sequence are 2048 long? Also this formula becomes wrong with CL activate

File: train/tr10-13B-ml/tr10-13B.slurm [hunk 4]
```diff
@@ -57,13 +57,14 @@ OPTIMIZER_ARGS=" \
     --adam-beta1 0.9 \
     --adam-beta2 0.95 \
     --adam-eps 1e-8 \
-    --lr 6e-5 \
+    --lr 1e-4 \
```
Comment (thomasw21) [inline train/tr10-13B-ml/tr10-13B.slurm:60]: GPT3 paper suggest a higher learning rate. Is there a reason why we would use `6e-5`?

**Review by thomasw21 — COMMENTED** [review]
Some thoughts on the config, I'll update the readme at the end so that we leave a trail with the conclusions we come up with.

File: train/tr10-13B-ml/tr10-13B.slurm (hunk 3 above)
Comment (stas00) [reply train/tr10-13B-ml/tr10-13B.slurm:84]: If I'm not mistaken you have mixed up `--train-tokens` here. This is samples, not tokens. 300M samples is ~600B tokens with seqlen 2048. So in the original it should have been `--train-samples 150_000_000` to match the 300B tokens, but it wasn't an issue as we stopped exactly at 150M samples.

Comment (stas00) [reply train/tr10-13B-ml/tr10-13B.slurm:84]: please let's not use this format, who can count all those zeros? it's too error-prone.

Comment (stas00) [reply train/tr10-13B-ml/tr10-13B.slurm:84]: I think the correct math is: `300_000_000_000/2048 = 146_484_375`

File: train/tr10-13B-ml/tr10-13B.slurm [hunk 5]
```diff
@@ -57,13 +57,14 @@ OPTIMIZER_ARGS=" \
...
     --min-lr 6e-6 \
     --lr-decay-style cosine \
-    --lr-decay-samples 126_953_125 \
```
Comment (stas00) [inline train/tr10-13B-ml/tr10-13B.slurm:63]: you removed this one w/o any commentary?

Comment (stas00) [reply train/tr10-13B-ml/tr10-13B.slurm:63]: The original tr1-13B said:

> We need lr-decay in samples, so tokens2samples = 260B / 2048 = 126_953_125

File: train/tr10-13B-ml/tr10-13B.slurm (hunk 3 above)
Comment (stas00) [reply train/tr10-13B-ml/tr10-13B.slurm:84]: I'd just set it at a round 150M

File: train/tr10-13B-ml/tr10-13B.slurm [hunk 6]
```diff
@@ -165,7 +166,7 @@ export CMD=" \
     --load $CHECKPOINT_PATH \
     --data-path $DATA_PATH \
     --data-impl mmap \
-    --split 900,100,0 \
+    --split 950,50,0 \
```
Comment (stas00) [inline train/tr10-13B-ml/tr10-13B.slurm:169]: currently using a small dataset, so I had to give valid a larger chunk. But for the real training this needs to be restored to the above split.

File: train/tr10-13B-ml/tr10-13B.slurm (hunk 1 above)
Comment (stas00) [reply train/tr10-13B-ml/tr10-13B.slurm:49]: FWIW, 530B training used:
```
NLAYERS=105
NHIDDEN=20480
NHEADS=128
```

So the same proportion as 32 and 5120

Comment (stas00) [reply train/tr10-13B-ml/tr10-13B.slurm:49]: Also, @TevenLeScao shared elsewhere a research paper showing that many heads we found to quite redundant anyway.

I'm not sure if there is a research showing size of the head vs. number of the heads performance.

File: train/tr10-13B-ml/tr10-13B.slurm (hunk 5 above)
Comment (thomasw21) [reply train/tr10-13B-ml/tr10-13B.slurm:63]: I was looking at setting it by default to the entire number of samples we have
https://github.com/bigscience-workshop/Megatron-DeepSpeed/blob/fd1e1da967c74e598acfc011031474663ef5845e/megatron/training.py#L341

We have been using this in arch/scaling. 

However I've just re-read the GPT3 paper and they do it for 260B ... so not sure here. cc @TevenLeScao

Comment (stas00) [reply train/tr10-13B-ml/tr10-13B.slurm:63]: Thank you for the note, Thomas - it's crucial that we leave a note trail, otherwise we have no idea why some config was added or removed.

File: train/tr10-13B-ml/tr10-13B.slurm [hunk 7]
```diff
@@ -81,7 +81,7 @@ GPT_ARGS=" \
     --micro-batch-size $MICRO_BATCH_SIZE \
     --rampup-batch-size 16 16 6_000_000 \
     --global-batch-size $GLOBAL_BATCH_SIZE \
-    --train-samples $((3000000000 / $SEQ_LEN + 1)) \
+    --train-samples $((300_000_000_000 / $SEQ_LEN + 1)) \
```
Comment (stas00) [inline train/tr10-13B-ml/tr10-13B.slurm:84]: except it doesn't work in bash:

```
bash: 300_000_000_000: value too great for base (error token is "300_000_000_000")
```

The easiest I found is to use Perl:

```
SEQ_LEN=2048
TRAIN_TOKENS=300_000_000_000
TRAIN_SAMPLES=$(perl -e "print $TRAIN_TOKENS / $SEQ_LEN")
[...]
--train-samples $TRAIN_SAMPLES
```

If you prefer python, so that everybody understands what exactly happens, then you have to add `int`:
```
python -c "print(int($TRAIN_TOKENS / $SEQ_LEN))"
```
either way works for me.

Comment (thomasw21) [reply train/tr10-13B-ml/tr10-13B.slurm:84]: I knew there was a reason I dropped that previous formatting ... thanks!

Comment (thomasw21) [reply train/tr10-13B-ml/tr10-13B.slurm:84]: actually you don't need to add `int` you should be able to run:
```
TRAIN_SAMPLES=$(python -c "print($TRAIN_TOKENS // $SEQ_LEN)")
```
`//` does euclidean division and returns the quotient.

Comment (stas00) [reply train/tr10-13B-ml/tr10-13B.slurm]: the new version looks great!
